annotation_pipeline/molecular_db.py

```diff
@@ -121,18 +121,35 @@ def _store(segm_i):
     segments_n = set().union(*pw.get_result(futures))
     append_pywren_stats(futures, pw.config['pywren']['runtime_memory'])
 
-    def deduplicate_formulas(segm_i, ibm_cos):
+    def deduplicate_formulas_segment(segm_i, ibm_cos, clean=True):
         objs = ibm_cos.list_objects_v2(Bucket=bucket, Prefix=f'{formulas_chunks_prefix}/chunk/{segm_i}/')
         keys = [obj['Key'] for obj in objs['Contents']]
 
         segm = set()
         for key in keys:
             segm_formulas_chunk = pickle.loads(ibm_cos.get_object(Bucket=bucket, Key=key)['Body'].read())
             segm.update(segm_formulas_chunk)
-        clean_from_cos(config, bucket, f'{formulas_chunks_prefix}/chunk/{segm_i}/', ibm_cos)
 
-        segm = pd.DataFrame(sorted(segm), columns=['formula'])
-        segm.index.name = 'formula_i'
+        if clean:
+            clean_from_cos(config, bucket, f'{formulas_chunks_prefix}/chunk/{segm_i}/', ibm_cos)
+
+        return segm
+
+    def get_formulas_number_per_chunk(segm_i, ibm_cos):
+        segm = deduplicate_formulas_segment(segm_i, ibm_cos, clean=False)
+        return len(segm)
+
+    pw = pywren.ibm_cf_executor(config=config, runtime_memory=2048)
+    futures = pw.map(get_formulas_number_per_chunk, segments_n)
+    formulas_nums = pw.get_result(futures)
+    append_pywren_stats(futures, pw.config['pywren']['runtime_memory'])
+
+    def store_formulas_segment(segm_i, ibm_cos):
+        segm = deduplicate_formulas_segment(segm_i, ibm_cos)
+        segm = pd.DataFrame(enumerate(sorted(segm)), columns=['formula_i', 'formula'])
+        segm['formula_i'] = segm['formula_i'].apply(lambda i: i + sum(formulas_nums[:segm_i]))
+        segm.set_index('formula_i', inplace=True)
```

Comment:
BTW, Pandas actually has an optimized type for sequential indexes:

```python
segm = deduplicate_formulas_segment(segm_i, ibm_cos)
formula_i_start = sum(formulas_nums[:segm_i])
formula_i_end = formula_i_start + len(segm)
segm = pd.DataFrame(sorted(segm),
                            columns=['formula'],
                            index=pd.RangeIndex(formula_i_start, formula_i_end, name='formula_i'))
```

`pd.RangeIndex` has the advantage that it doesn't need to store a value for each row (it just stores one set of `start, end, step` for the whole table), so `segm` uses ~33% less space.